### PR TITLE
mTLS authn method

### LINKED ISSUE #8: mTLS authN mode
There's already a placeholder for it at https://github.com/3scale-labs/authorino/blob/2b6a6f8016a5837650506cd125751f8d95ea4197/pkg/config/identity/mtls.go.

### PR DESCRIPTION
**Pull request by guicassolato**
Adds support for client TLS certificate validation at application layer in Authorino for connections with the protected service and Authorino raw HTTP authorization interface relying on Mutual Transport Layer Security (mTLS) for authentication.

Closes #8

### Verification steps

Create a cluster, build and deploy the apps:

```sh
make local-setup FF=1
```

Create a CA certificate and store it in a secret to be discovered by Authorino:

```sh
openssl req -x509 -sha256 -days 365 -nodes -newkey rsa:2048 -subj "/CN=talker-api-ca" -keyout /tmp/ca.key -out /tmp/ca.crt
kubectl create secret tls talker-api-ca --cert=/tmp/ca.crt --key=/tmp/ca.key
kubectl label secret talker-api-ca authorino.kuadrant.io/managed-by=authorino app=talker-api
```

Redeploy Envoy acknowledging the CA cert and enabling client certificate validation:

(This steps sucks, but otherwise Envoy will not pass the client peer certificate to Authorino in the payload to external authorization. Effectively, we're adding redundant client cert validation – i.e. Envoy doing it at L4 and Authorino at L7. For a hope on finding a way to avoid this, see https://github.com/envoyproxy/envoy/issues/20563#issuecomment-1152242226.)

```sh
kubectl delete deployment/envoy
kubectl apply -f -<<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app: envoy
  name: envoy
data:
  envoy.yaml: |
    static_resources:
      listeners:
      - address:
          socket_address:
            address: 0.0.0.0
            port_value: 8000
        filter_chains:
        - transport_socket:
            name: envoy.transport_sockets.tls
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
              common_tls_context:
                tls_certificates:
                - certificate_chain: {filename: "/etc/ssl/certs/talker-api/tls.crt"}
                  private_key: {filename: "/etc/ssl/certs/talker-api/tls.key"}
                validation_context:
                  trusted_ca:
                    filename: /etc/ssl/certs/talker-api/tls.crt
          filters:
          - name: envoy.http_connection_manager
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
              stat_prefix: local
              route_config:
                name: local_route
                virtual_hosts:
                - name: local_service
                  domains: ['*']
                  routes:
                  - match: { prefix: / }
                    route: { cluster: talker-api }
              http_filters:
              - name: envoy.filters.http.ext_authz
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
                  transport_api_version: V3
                  failure_mode_allow: false
                  include_peer_certificate: true
                  grpc_service:
                    envoy_grpc: { cluster_name: authorino }
                    timeout: 1s
              - name: envoy.filters.http.router
                typed_config: {}
              use_remote_address: true
      clusters:
      - name: authorino
        connect_timeout: 0.25s
        type: strict_dns
        lb_policy: round_robin
        http2_protocol_options: {}
        load_assignment:
          cluster_name: authorino
          endpoints:
          - lb_endpoints:
            - endpoint:
                address:
                  socket_address:
                    address: authorino-authorino-authorization
                    port_value: 50051
        transport_socket:
          name: envoy.transport_sockets.tls
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
            common_tls_context:
              validation_context:
                trusted_ca:
                  filename: /etc/ssl/certs/authorino-ca-cert.crt
      - name: talker-api
        connect_timeout: 0.25s
        type: strict_dns
        lb_policy: round_robin
        load_assignment:
          cluster_name: talker-api
          endpoints:
          - lb_endpoints:
            - endpoint:
                address:
                  socket_address:
                    address: talker-api
                    port_value: 3000
    admin:
      access_log_path: "/tmp/admin_access.log"
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 8001
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: envoy
  name: envoy
spec:
  selector:
    matchLabels:
      app: envoy
  template:
    metadata:
      labels:
        app: envoy
    spec:
      containers:
      - args:
        - --config-path /usr/local/etc/envoy/envoy.yaml
        - --service-cluster front-proxy
        - --log-level info
        - --component-log-level filter:trace,http:debug,router:debug
        command:
        - /usr/local/bin/envoy
        image: envoyproxy/envoy:v1.19-latest
        name: envoy
        ports:
        - containerPort: 8000
          name: web
        - containerPort: 8001
          name: admin
        volumeMounts:
        - mountPath: /usr/local/etc/envoy
          name: config
          readOnly: true
        - mountPath: /etc/ssl/certs/authorino-ca-cert.crt
          name: authorino-ca-cert
          readOnly: true
          subPath: ca.crt
        - mountPath: /etc/ssl/certs/talker-api
          name: talker-api-ca
          readOnly: true
      volumes:
      - configMap:
          items:
          - key: envoy.yaml
            path: envoy.yaml
          name: envoy
        name: config
      - name: authorino-ca-cert
        secret:
          defaultMode: 420
          secretName: authorino-ca-cert
      - name: talker-api-ca
        secret:
          defaultMode: 420
          secretName: talker-api-ca
---
apiVersion: v1
kind: Service
metadata:
  name: envoy
spec:
  selector:
    app: envoy
  ports:
  - name: web
    port: 8000
    protocol: TCP
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ingress-wildcard-host
spec:
  rules:
  - host: talker-api-authorino.127.0.0.1.nip.io
    http:
      paths:
      - backend:
          service:
            name: envoy
            port: { number: 8000 }
        path: /
        pathType: Prefix
EOF
```

Create the AuthConfig:

```sh
kubectl apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: talker-api-protection
spec:
  hosts:
  - talker-api-authorino.127.0.0.1.nip.io
  identity:
  - name: mtls
    mtls:
      labelSelectors:
        app: talker-api
  authorization:
  - name: acme
    json:
      rules:
      - selector: auth.identity.Organization
        operator: incl
        value: ACME Inc.
EOF
```

Try the API with a TLS certificate signed by the trusted CA:

```sh
openssl genrsa -out /tmp/aisha.key 2048
openssl req -new -key /tmp/aisha.key -out /tmp/aisha.csr -subj "/CN=aisha/C=PK/L=Islamabad/O=ACME Inc./OU=Engineering"
openssl x509 -req -in /tmp/aisha.csr -CA /tmp/ca.crt -CAkey /tmp/ca.key -CAcreateserial -out /tmp/aisha.crt -days 1 -sha256

curl -k --cert /tmp/aisha.crt --key /tmp/aisha.key https://talker-api-authorino.127.0.0.1.nip.io:8000 -i
# HTTP/1.1 200 OK
```

Try the API with a TLS certificate signed by the trusted CA, though missing an authorized Organization:

```sh
openssl genrsa -out /tmp/john.key 2048
openssl req -new -key /tmp/john.key -out /tmp/john.csr -subj "/CN=john/C=UK/L=London"
openssl x509 -req -in /tmp/john.csr -CA /tmp/ca.crt -CAkey /tmp/ca.key -CAcreateserial -out /tmp/john.crt -days 1 -sha256

curl -k --cert /tmp/john.crt --key /tmp/john.key https://talker-api-authorino.127.0.0.1.nip.io:8000 -i
# HTTP/1.1 403 Forbidden
# x-ext-auth-reason: Unauthorized
```

Try the AuthConfig via raw HTTP authorization interface:

Expose the interface:

```sh
kubectl port-forward service/authorino-authorino-authorization 5001:5001 &
```

Consume the raw HTTP authorization interface with a TLS certificate signed by the trusted CA:

```sh
curl -k --cert /tmp/aisha.crt --key /tmp/aisha.key -H 'Content-Type: application/json' -d '{}' https://talker-api-authorino.127.0.0.1.nip.io:5001/check -i
# HTTP/2 200
```

Consume the raw HTTP authorization interface with a TLS certificate signed by a unknown authority:

```sh
openssl req -x509 -sha256 -days 365 -nodes -newkey rsa:2048 -subj "/CN=untrusted" -keyout /tmp/untrusted-ca.key -out /tmp/untrusted-ca.crt
openssl genrsa -out /tmp/niko.key 2048
openssl req -new -key /tmp/niko.key -out /tmp/niko.csr -subj "/CN=niko/C=JP/L=Osaka"
openssl x509 -req -in /tmp/niko.csr -CA /tmp/untrusted-ca.crt -CAkey /tmp/untrusted-ca.key -CAcreateserial -out /tmp/niko.crt -days 1 -sha256

curl -k --cert /tmp/niko.crt --key /tmp/niko.key -H 'Content-Type: application/json' -d '{}' https://talker-api-authorino.127.0.0.1.nip.io:5001/check -i
# HTTP/2 401
# www-authenticate: Basic realm="mtls"
# x-ext-auth-reason: x509: certificate signed by unknown authority
```

Revoke an entire chain of certificates:

```sh
kubectl delete secret/talker-api-ca
```

Even if the deleted root certificate is still cached and accepted at the gateway, Authorino will revoke access at application level immediately.

Try the API again with a previously accepted certificate to verify:

```sh
curl -k --cert /tmp/aisha.crt --key /tmp/aisha.key https://talker-api-authorino.127.0.0.1.nip.io:8000 -i
# HTTP/1.1 401 Unauthorized
# www-authenticate: Basic realm="mtls"
# x-ext-auth-reason: x509: certificate signed by unknown authority
```

Cleanup:

```sh
make local-cleanup
```

### CODE REVIEW

File: controllers/secret_controller.go [hunk 1]
```diff
@@ -79,12 +73,18 @@ func (r *SecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-func (r *SecretReconciler) getAuthConfigsUsingAPIKey(ctx context.Context) authConfigSet {
+func (r *SecretReconciler) eachAuthConfigsWithK8sSecretBasedIdentity(ctx context.Context, f func(*evaluators.AuthConfig)) {
```
Comment (alexsnaps) [inline controllers/secret_controller.go:76]: We don't need passing the `ctx context.Context` in here, do we?
Unclear to me what our stance here is/should be. I've seen a few patterns in the code (first arg, last arg, pass it despite not being used, not pass it, ...)

**Comment by guicassolato** [comment]
### Verification steps using intermediate CAs

The certificate tree in this example looks like the following:

cert-tree

Steps:

```sh
# build and deploy everything
make local-setup FF=1

# create the authconfig
kubectl apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: talker-api-protection
spec:
  hosts:
  - talker-api-authorino.127.0.0.1.nip.io
  identity:
  - name: mtls
    mtls:
      labelSelectors:
        app: talker-api
  authorization:
  - name: acme
    json:
      rules:
      - selector: auth.identity.Organization
        operator: incl
        value: ACME Inc.
EOF

# create the root ca and store it in a k8s secret
mkdir /tmp/mtls

cat <<EOF > /tmp/mtls/openssl.cnf
HOME                    = .
RANDFILE                = $ENV::HOME/.rnd
oid_section             = new_oids

[ new_oids ]

[ ca ]
default_ca              = CA_default

[ CA_default ]
dir                     = /tmp/mtls
new_certs_dir           = /tmp/mtls
database                = /tmp/mtls/index.txt
serial                  = /tmp/mtls/serial
certificate             = /tmp/mtls/ca.crt
private_key             = /tmp/mtls/ca.key
default_md              = sha256 
policy                  = policy_match

[ policy_match ]
countryName             = optional
stateOrProvinceName     = optional
organizationName        = optional
organizationalUnitName  = optional
commonName              = supplied
emailAddress            = optional

[ policy_anything ]
countryName             = optional
stateOrProvinceName     = optional
localityName            = optional
organizationName        = optional
organizationalUnitName  = optional
commonName              = supplied
emailAddress            = optional

[ req ]
default_bits            = 4096
default_md              = sha256
default_keyfile         = privkey.pem
distinguished_name      = req_distinguished_name
x509_extensions         = v3_ca
string_mask             = nombstr

[ req_distinguished_name ]
countryName                     = Country Name (2 letter code)
countryName_default             = ES
countryName_min                 = 2
countryName_max                 = 2
stateOrProvinceName             = State or Province Name (full name)
stateOrProvinceName_default     = Barcelona
localityName                    = Locality Name (eg, city)
localityName_default            = Barcelona
0.organizationName              = Organization Name (eg, company)
0.organizationName_default      = ACME Inc.
organizationalUnitName          = Organizational Unit Name (eg, section)
commonName                      = Common Name (eg, your name or your server\'s hostname)
commonName_max                  = 64
emailAddress                    = Email Address
emailAddress_max                = 64

[ v3_ca ]
subjectKeyIdentifier=hash
authorityKeyIdentifier=keyid:always,issuer
basicConstraints = critical,CA:true

[ v3_intermediate_ca ]
subjectKeyIdentifier = hash
authorityKeyIdentifier = keyid:always,issuer
basicConstraints = critical, CA:true, pathlen:0
keyUsage = critical, digitalSignature, cRLSign, keyCertSign
EOF

echo 01 > /tmp/mtls/serial
touch /tmp/mtls/index.txt

openssl req -x509 -sha256 -days 3650 -nodes -newkey rsa:2048 -config /tmp/mtls/openssl.cnf -extensions v3_ca -subj "/CN=talker-api-authorino.127.0.0.1.nip.io" -keyout /tmp/mtls/ca.key -out /tmp/mtls/ca.crt
kubectl create secret tls talker-api-ca --cert=/tmp/mtls/ca.crt --key=/tmp/mtls/ca.key

# redeploy envoy to start using the root ca
kubectl delete deployment/envoy
kubectl apply -f -<<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app: envoy
  name: envoy
data:
  envoy.yaml: |
    static_resources:
      listeners:
      - address:
          socket_address:
            address: 0.0.0.0
            port_value: 8000
        filter_chains:
        - transport_socket:
            name: envoy.transport_sockets.tls
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
              common_tls_context:
                tls_certificates:
                - certificate_chain: {filename: "/etc/ssl/certs/talker-api/tls.crt"}
                  private_key: {filename: "/etc/ssl/certs/talker-api/tls.key"}
                validation_context:
                  trusted_ca:
                    filename: /etc/ssl/certs/talker-api/tls.crt
          filters:
          - name: envoy.http_connection_manager
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
              stat_prefix: local
              route_config:
                name: local_route
                virtual_hosts:
                - name: local_service
                  domains: ['*']
                  routes:
                  - match: { prefix: / }
                    route: { cluster: talker-api }
              http_filters:
              - name: envoy.filters.http.ext_authz
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
                  transport_api_version: V3
                  failure_mode_allow: false
                  include_peer_certificate: true
                  grpc_service:
                    envoy_grpc: { cluster_name: authorino }
                    timeout: 1s
              - name: envoy.filters.http.router
                typed_config: {}
              use_remote_address: true
      clusters:
      - name: authorino
        connect_timeout: 0.25s
        type: strict_dns
        lb_policy: round_robin
        http2_protocol_options: {}
        load_assignment:
          cluster_name: authorino
          endpoints:
          - lb_endpoints:
            - endpoint:
                address:
                  socket_address:
                    address: authorino-authorino-authorization
                    port_value: 50051
        transport_socket:
          name: envoy.transport_sockets.tls
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
            common_tls_context:
              validation_context:
                trusted_ca:
                  filename: /etc/ssl/certs/authorino-ca-cert.crt
      - name: talker-api
        connect_timeout: 0.25s
        type: strict_dns
        lb_policy: round_robin
        load_assignment:
          cluster_name: talker-api
          endpoints:
          - lb_endpoints:
            - endpoint:
                address:
                  socket_address:
                    address: talker-api
                    port_value: 3000
    admin:
      access_log_path: "/tmp/admin_access.log"
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 8001
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: envoy
  name: envoy
spec:
  selector:
    matchLabels:
      app: envoy
  template:
    metadata:
      labels:
        app: envoy
    spec:
      containers:
      - args:
        - --config-path /usr/local/etc/envoy/envoy.yaml
        - --service-cluster front-proxy
        - --log-level info
        - --component-log-level filter:trace,http:debug,router:debug
        command:
        - /usr/local/bin/envoy
        image: envoyproxy/envoy:v1.19-latest
        name: envoy
        ports:
        - containerPort: 8000
          name: web
        - containerPort: 8001
          name: admin
        volumeMounts:
        - mountPath: /usr/local/etc/envoy
          name: config
          readOnly: true
        - mountPath: /etc/ssl/certs/authorino-ca-cert.crt
          name: authorino-ca-cert
          readOnly: true
          subPath: ca.crt
        - mountPath: /etc/ssl/certs/talker-api
          name: talker-api-ca
          readOnly: true
      volumes:
      - configMap:
          items:
          - key: envoy.yaml
            path: envoy.yaml
          name: envoy
        name: config
      - name: authorino-ca-cert
        secret:
          defaultMode: 420
          secretName: authorino-ca-cert
      - name: talker-api-ca
        secret:
          defaultMode: 420
          secretName: talker-api-ca
---
apiVersion: v1
kind: Service
metadata:
  name: envoy
spec:
  selector:
    app: envoy
  ports:
  - name: web
    port: 8000
    protocol: TCP
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ingress-wildcard-host
spec:
  rules:
  - host: talker-api-authorino.127.0.0.1.nip.io
    http:
      paths:
      - backend:
          service:
            name: envoy
            port: { number: 8000 }
        path: /
        pathType: Prefix
EOF
kubectl port-forward deployment/envoy 8000:8000 &

# create 3 intermediate cas and store 2 of them in k8s secrets watched by authorino
openssl genrsa -out /tmp/mtls/intermediate-ca-1.key 2048
openssl req -new -config /tmp/mtls/openssl.cnf -key /tmp/mtls/intermediate-ca-1.key -out /tmp/mtls/intermediate-ca-1.csr -subj "/CN=intermediate-1"
openssl ca -config /tmp/mtls/openssl.cnf -in /tmp/mtls/intermediate-ca-1.csr -out /tmp/mtls/intermediate-ca-1.crt -extensions v3_intermediate_ca -days 3650
openssl x509 -in /tmp/mtls/intermediate-ca-1.crt -out /tmp/mtls/intermediate-ca-1.crt -outform PEM
kubectl create secret tls intermediate-ca-1 --cert=/tmp/mtls/intermediate-ca-1.crt --key=/tmp/mtls/intermediate-ca-1.key
kubectl label secret intermediate-ca-1 authorino.kuadrant.io/managed-by=authorino app=talker-api

openssl genrsa -out /tmp/mtls/intermediate-ca-2.key 2048
openssl req -new -config /tmp/mtls/openssl.cnf -key /tmp/mtls/intermediate-ca-2.key -out /tmp/mtls/intermediate-ca-2.csr -subj "/CN=intermediate-2"
openssl ca -config /tmp/mtls/openssl.cnf -in /tmp/mtls/intermediate-ca-2.csr -out /tmp/mtls/intermediate-ca-2.crt -extensions v3_intermediate_ca -days 3650
openssl x509 -in /tmp/mtls/intermediate-ca-2.crt -out /tmp/mtls/intermediate-ca-2.crt -outform PEM
kubectl create secret tls intermediate-ca-2 --cert=/tmp/mtls/intermediate-ca-2.crt --key=/tmp/mtls/intermediate-ca-2.key
kubectl label secret intermediate-ca-2 authorino.kuadrant.io/managed-by=authorino app=talker-api

openssl genrsa -out /tmp/mtls/intermediate-ca-3.key 2048
openssl req -new -config /tmp/mtls/openssl.cnf -key /tmp/mtls/intermediate-ca-3.key -out /tmp/mtls/intermediate-ca-3.csr -subj "/CN=intermediate-3"
openssl ca -config /tmp/mtls/openssl.cnf -in /tmp/mtls/intermediate-ca-3.csr -out /tmp/mtls/intermediate-ca-3.crt -extensions v3_intermediate_ca -days 3650
openssl x509 -in /tmp/mtls/intermediate-ca-3.crt -out /tmp/mtls/intermediate-ca-3.crt -outform PEM

# generate leaf certificates signed with each intermediate ca and try to authenticate to the api with them
openssl genrsa -out /tmp/mtls/aisha.key 2048
openssl req -new -config /tmp/mtls/openssl.cnf -key /tmp/mtls/aisha.key -out /tmp/mtls/aisha.csr -subj "/CN=aisha/C=PK/L=Islamabad/O=ACME Inc./OU=Engineering"
openssl x509 -req -in /tmp/mtls/aisha.csr -out /tmp/mtls/aisha.crt -CA /tmp/mtls/intermediate-ca-1.crt -CAkey /tmp/mtls/intermediate-ca-1.key -CAcreateserial -days 1 -sha256
cat /tmp/mtls/intermediate-ca-1.crt >> /tmp/mtls/aisha.crt

curl -k --cert /tmp/mtls/aisha.crt --key /tmp/mtls/aisha.key https://talker-api-authorino.127.0.0.1.nip.io:8000 -i
# 200

openssl genrsa -out /tmp/mtls/john.key 2048
openssl req -new -config /tmp/mtls/openssl.cnf -key /tmp/mtls/john.key -out /tmp/mtls/john.csr -subj "/CN=john/C=UK/L=London"
openssl x509 -req -in /tmp/mtls/john.csr -out /tmp/mtls/john.crt -CA /tmp/mtls/intermediate-ca-2.crt -CAkey /tmp/mtls/intermediate-ca-2.key -CAcreateserial -days 1 -sha256
cat /tmp/mtls/intermediate-ca-2.crt >> /tmp/mtls/john.crt

curl -k --cert /tmp/mtls/john.crt --key /tmp/mtls/john.key https://talker-api-authorino.127.0.0.1.nip.io:8000 -i
# 403

openssl genrsa -out /tmp/mtls/niko.key 2048
openssl req -new -config /tmp/mtls/openssl.cnf -key /tmp/mtls/niko.key -out /tmp/mtls/niko.csr -subj "/CN=niko/C=JP/L=Osaka"
openssl x509 -req -in /tmp/mtls/niko.csr -out /tmp/mtls/niko.crt -CA /tmp/mtls/intermediate-ca-3.crt -CAkey /tmp/mtls/intermediate-ca-3.key -CAcreateserial -days 1 -sha256
cat /tmp/mtls/intermediate-ca-3.crt >> /tmp/mtls/niko.crt

curl -k --cert /tmp/mtls/niko.crt --key /tmp/mtls/niko.key https://talker-api-authorino.127.0.0.1.nip.io:8000 -i
# 401
```

**Review by alexsnaps — APPROVED** [review]
There is nothing that jumps out at me other than a few general remarks:

 - I _think_ convention is to pass a `ctx` as the first argument of a method. But, even if do this differently (I honestly don't care), I think it'd be nice to have some consistency.
 - What heuristic do we use for even passing a `ctx`, if we don't need it. Mine has been use it on interfaces, but otherwise leave it out.
 - I try to keep changes to a minimum, e.g. renaming receiver of methods or import identifiers. 

Just so to make it easier for people to review the changes. Not sure how we'd rather work on this code base? Some people like doing refactoring as complete different PRs (as to rename receiver of method for instance), or just keep them as a separate commit in one PR, so the diff refactoring vs functional diffs can be reviewed independently... 

Anyways, just wanted to share what I found making reviews easier. That being said I know how easy to just oversee these things when in the zone implementing something :)

File: controllers/secret_controller.go (hunk 1 above)
Comment (guicassolato) [reply controllers/secret_controller.go:76]: Well spotted! I think I left this one behind in the cleanup. I was using it for debugging logs added temporarily in dev, that later didn't make sense to keep in the final version.

Comment (guicassolato) [reply controllers/secret_controller.go]: Fixed

**Comment by guicassolato** [comment]
Thanks for the review, @alexsnaps!

> I think convention is to pass a ctx as the first argument of a method. But, even if do this differently (I honestly don't care), I think it'd be nice to have some consistency.

I think you're referring to `AuthConfigEvaluator.Call(AuthPipeline, context.Context)`, right? Sorry. Bad decisions in the past.

> I try to keep changes to a minimum, e.g. renaming receiver of methods or import identifiers.

> Some people like doing refactoring as complete different PRs (as to rename receiver of method for instance), or just keep them as a separate commit in one PR, so the diff refactoring vs functional diffs can be reviewed independently...

We can still extract https://github.com/Kuadrant/authorino/pull/305/commits/f7660c58a6a4ee18a030c257414567dd3cc4998d and https://github.com/Kuadrant/authorino/pull/305/commits/5349997f9c0f43b94ad956f4961cf7f8a2d5afd3 to a separate PR if you think it's worth it. I guess I was just lazy not to do so in the first place, hoping people would review the history.

Same does not apply to renaming `decodeCertificate -> certificateFromSecret` in https://github.com/Kuadrant/authorino/pull/305/commits/2c59f8805e8cbf6cc560e883419b94a933f7ade9. That was rubbish. Sincere apologies.

File: install/manifests.yaml [hunk 2]
```diff
@@ -916,6 +923,25 @@ spec:
                       description: Whether this identity config should generate individual
                         observability metrics
                       type: boolean
+                    mtls:
+                      properties:
+                        allNamespaces:
+                          default: false
+                          description: Whether Authorino should look for TLS secrets
+                            in all namespaces or only in the same namespace of the
+                            AuthConfig. Enabling this option in namespaced Authorino
```
Comment (alexsnaps) [inline install/manifests.yaml:932]: `in the same namespace as the AuthConfig` or `in the namespace of the AuthConfig` ?

Comment (guicassolato) [reply install/manifests.yaml:932]: Fixed. Thanks.

**Pull request merged**